presto-main/src/test/java/io/prestosql/sql/query/TestHaving.java

```diff
@@ -13,18 +13,58 @@
  */
 package io.prestosql.sql.query;
 
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
 public class TestHaving
 {
+    private QueryAssertions assertions;
+
+    @BeforeClass
+    public void init()
+    {
+        assertions = new QueryAssertions();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void teardown()
+    {
+        assertions.close();
+        assertions = null;
+    }
+
     @Test
     public void testImplicitGroupBy()
     {
-        try (QueryAssertions queryAssertions = new QueryAssertions()) {
-            queryAssertions
-                    .assertQuery(
-                            "SELECT 'x' FROM (VALUES 1, 1, 2) t(a) HAVING true",
-                            "VALUES 'x'");
-        }
+        assertions.assertQuery(
+                "SELECT 'x' FROM (VALUES 1, 1, 2) t(a) HAVING true",
+                "VALUES 'x'");
+    }
+
+    @Test
+    public void testCorrelation()
+    {
+        // unqualified reference
+        assertions.assertQuery(
```

Comment:
Could you also test some unsupported correlated aggregations, e.g
```

                "SELECT * " +
                        "FROM (VALUES 1) t(x) " +
                        "WHERE EXISTS (" +
                        "    SELECT count(x)" +
                        "    FROM (VALUES 1, 2) u(y)" +
                        "    GROUP BY y" +
                        "    HAVING y = x)",
```
and see if error is reasonable

Reply:
Why would that be unsupported?

Reply:
Constant in `count(x)` won't be resolved.

Reply:
Well, it does work, actually :)

Reply:
It works in master if the HAVING clause is removed, too.

Reply:
I think it works because there is `y = x` condition which makes subquery rewritten to `count(y)`, so decorrelation kicks-in